akka-actor-typed/src/main/scala/akka/actor/typed/delivery/WorkPullingProducerController.scala

```diff
@@ -0,0 +1,215 @@
+ *
+ * The producer and `WorkPullingProducerController` actors are supposed to be local so that these messages are
+ * fast and not lost.
+ *
+ * Many unconfirmed messages can be in flight between the `WorkPullingProducerController` and each
+ * `ConsumerController`. The flow control is driven by the consumer side, which means that the
+ * `WorkPullingProducerController` will not send faster than the demand requested by the workers.
+ *
+ * Lost messages are detected, resent and deduplicated if needed. This is also driven by the consumer side,
+ * which means that the `WorkPullingProducerController` will not push resends unless requested by the
+ * `ConsumerController`.
+ *
+ * If a worker crashes or is stopped gracefully the unconfirmed messages for that worker will be
+ * routed to other workers by the `WorkPullingProducerController`. This may result in that some messages
+ * may be processed more than once, by different workers.
+ *
+ * Until sent messages have been confirmed the `WorkPullingProducerController` keeps them in memory to be able to
+ * resend them. If the JVM of the `WorkPullingProducerController` crashes those unconfirmed messages are lost.
+ * To make sure the messages can be delivered also in that scenario the `WorkPullingProducerController` can be
+ * used with a [[DurableProducerQueue]]. Then the unconfirmed messages are stored in a durable way so
+ * that they can be redelivered when the producer is started again. An implementation of the
+ * `DurableProducerQueue` is provided by `EventSourcedProducerQueue` in `akka-persistence-typed`.
+ *
+ * Instead of using `tell` with the `sendNextTo` in the `RequestNext` the producer can use `context.ask`
+ * with the `askNextTo` in the `RequestNext`. The difference is that a reply is sent back when the
+ * message has been handled. If a `DurableProducerQueue` is used then the reply is sent when the message
+ * has been stored successfully, but it might not have been processed by the consumer yet. Otherwise the
+ * reply is sent after the consumer has processed and confirmed the message.
+ *
+ * It's also possible to use the `WorkPullingProducerController` and `ConsumerController` without resending
+ * lost messages, but the flow control is still used. This can for example be useful when both consumer and
+ * producer are know to be located in the same local `ActorSystem`. This can be more efficient since messages
+ * don't have to be kept in memory in the `ProducerController` until they have been
+ * confirmed, but the drawback is that lost messages will not be delivered. See configuration
+ * `only-flow-control` of the `ConsumerController`.
+ */
+@ApiMayChange // TODO when removing ApiMayChange consider removing `case class` for some of the messages
+object WorkPullingProducerController {
+
+  import WorkPullingProducerControllerImpl.UnsealedInternalCommand
+
+  sealed trait Command[A] extends UnsealedInternalCommand
```

Comment:
Pity to duplicated the protocol but I guess it stops accidnets when using multiple types of producer controllers

Reply:
they are also slightly different